peer/ping_manager.go

```diff
@@ -0,0 +1,259 @@
+package peer
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// pingManager is a structure that is designed to manage the internal state
+// of the ping pong lifecycle with the remote peer. We assume there is only one
+// ping outstanding at once.
+//
+// NOTE: This structure MUST be initialized with NewPingManager.
+type PingManager struct {
+	// newPingPayload is a closure that returns the payload to be packaged
+	// in the Ping message.
+	newPingPayload func() []byte
+
+	// newPongSize is a closure that returns a random value between
+	// [0, lnwire.MaxPongBytes]. This random value helps to more effectively
+	// pair Pong messages with Ping.
+	newPongSize func() uint16
+
+	// intervalDuration is the Duration between attempted pings.
+	intervalDuration time.Duration
+
+	// timeoutDuration is the Duration we wait before declaring a ping
+	// attempt failed.
+	timeoutDuration time.Duration
+
+	// pingTime is a rough estimate of the RTT (round-trip-time) between us
+	// and the connected peer.
+	// To be used atomically.
+	// TODO(roasbeef): also use a WMA or EMA?
+	pingTime atomic.Pointer[time.Duration]
+
+	// pingLastSend is the time when we sent our last ping message.
+	// To be used atomically.
+	pingLastSend *time.Time
+
+	// outstandingPongSize is the current size of the requested pong
+	// payload.  This value can only validly range from [0,65531]. Any
+	// value < 0 is interpreted as if there is no outstanding ping message.
+	outstandingPongSize int32
+
+	// pingTicker is a pointer to a Ticker that fires on every ping
+	// interval.
+	pingTicker *time.Ticker
+
+	// pingTimeout is a Timer that will fire when we want to time out a
+	// ping
+	pingTimeout *time.Timer
+
+	// pingChan is the channel on which the pingManager will write Ping
+	// messages it wishes to send out
+	pingChan chan<- lnwire.Ping
+
+	// pongChan is the channel on which the pingManager will write Pong
+	// messages it is evaluating
+	pongChan <-chan lnwire.Pong
+
+	// failChan is the channel on which the pingManager will report ping
+	// failures. This will happen if the received Pong message deviates
+	// from what is acceptable, or if it times out.
+	failChan chan<- struct{}
```

Comment:
Why? that would be a destructive change since closing a channel twice would blow up. Instead this construction allows the surrounding context to decide what to do if the failure occurs and there is no clear reason why we should be forced to disconnect if a pong failure occurs. This code separates concerns more effectively than if we were to close the channel imo.

Reply:
Originally thought this was performing the job of the quit channel. have removed this in my patch in any case.